### ISSUE TITLE
decompress-point:secp256k1:secp:crypt doesn't use the compressed key

### ISSUE THREAD

**Issue by yosoyubik** [urbit/urbit#2877]
**Describe the bug**
`decompress-point:secp256k1:secp:crypt` uses `a` (from `y^2=x^3+ax+b`, which is zero for secp256k1) instead of the compressed key (`dat`) [[source]](https://github.com/urbit/urbit/blob/master/pkg/arvo/sys/zuse.hoon#L4389) to calculate the x coordinate of the curve point.

**To Reproduce**
Steps to reproduce the behaviour:
```
> (decompress-point:secp256k1:secp:crypto 0x3.57bf.e1e3.41d0.1c69.fe56.5430.9956.cbea.5168.22fb.a8a6.0174.3a01.2a78.96ee.8dc5)
[x=0 y=7]
> (decompress-point:secp256k1:secp:crypto 0x3.57bf.e1e3.41d0.1c69.fe56.5430.9956.cbea.5168.22fb.a8a6.0174.3a01.2a78.96ee.8dc1)
[x=0 y=7]
> (decompress-point:secp256k1:secp:crypto 0x3.57bf.e1e3.41d0.1c69.fe56.5430.9956.cbea.5168.22fb.a8a6.0174.3a01.2a78.96ee.8dc2)
[x=0 y=7]
> (decompress-point:secp256k1:secp:crypto 0x3.57bf.e1e3.41d0.1c69.fe56.5430.9956.cbea.5168.22fb.a8a6.0174.3a01.2a78.96ee.8dc3)
[x=0 y=7]
> (decompress-point:secp256k1:secp:crypto 0x3.57bf.e1e3.41d0.1c69.fe56.5430.9956.cbea.5168.22fb.a8a6.0174.3a01.2a78.96ee.8dc4)
[x=0 y=7]
```

**Expected behaviour**
```
> (decompress-point:secp256k1:secp:crypto 0x3.57bf.e1e3.41d0.1c69.fe56.5430.9956.cbea.5168.22fb.a8a6.0174.3a01.2a78.96ee.8dc1)
[ x=39.690.244.635.883.679.370.511.189.826.325.366.723.996.909.330.137.750.917.244.362.082.889.677.835.713
    y
\/62.524.658.337.986.403.563.241.993.582.420.062.061.542.569.724.895.835.669.102.787.698.654.579.596.680.123.236.509.538.293.335.781.509.132.095.396.190.952.736.188.21\/
0.356.360.700.198.802.518.720.431.010.334.480.614.671.983.109.091.373.538.065.758.065.666.452.151.889.297.001.838.337.218.910.143.540.812.104
  \/                                                                                                                                                                     \/
]
> (decompress-point:secp256k1:secp:crypto 0x3.57bf.e1e3.41d0.1c69.fe56.5430.9956.cbea.5168.22fb.a8a6.0174.3a01.2a78.96ee.8dc2)
[ x=39.690.244.635.883.679.370.511.189.826.325.366.723.996.909.330.137.750.917.244.362.082.889.677.835.714
    y
\/62.524.658.337.986.403.563.241.993.582.420.062.061.542.569.724.895.835.669.102.787.698.654.579.596.684.849.183.067.307.172.887.825.803.375.055.659.159.364.740.927.99\/
8.543.225.720.239.138.993.100.520.839.667.109.663.856.181.566.838.331.178.353.469.680.699.061.566.506.502.191.355.146.066.031.716.028.974.351
  \/                                                                                                                                                                     \/
]
... etc
```

**Solution**
```
=+  x=(end 3 w dat)
```

**Notify maintainers**

@belisarius222 @Fang- @joemfb 

**Post Scriptum**

I noticed that `x` and `y` have different lengths after the solution was implemented. I don't know if this is wrong per se, but at least it's causing an issue when I derive a new public key from the one coming out of the decompress gate. 

I see that `y` is calculated without using modular operation ([over p](https://github.com/urbit/urbit/blob/master/pkg/arvo/sys/zuse.hoon#L4352)) which seems to be the way that at least is done [here](https://cryptobook.nakov.com/asymmetric-key-ciphers/elliptic-curve-cryptography-ecc#elliptic-curves-over-finite-fields). Other implementations I've seen use also modular arithmetic to obtain `y`.

I have a specific example using another implementation of secp256k1 ([Bcrypto](https://github.com/bcoin-org/bcrypto/blob/73dff63d64fd5ba9446780b0295b6a42b9adc536/lib/js/elliptic.js)). If my key is this:

```
> =key 0x357bfe1e341d01c69fe5654309956cbea516822fba8a601743a012a7896ee8dc2
```
x and y should be:
```
[
x=57bfe1e341d01c69fe5654309956cbea516822fba8a601743a012a7896ee8dc2 y=4310ef3676384179e713be3115e93f34ac9a3933f6367aeb3081527ea74027b7
]
```
I believe that this is correct because when I use this values for key derivation, I obtained the expected public key.

I get the correct result for x:

```
> x=`@x`(end 3 32 key)
x=0x57bfe1e341d01c69fe5654309956cbea516822fba8a601743a012a7896ee8dc2

```

I've tried with a number of modification of the calculation of `y` using modulo arithmetic without success. Some (wrong) examples here:

```
> =prim 0xffff.ffff.ffff.ffff.ffff.ffff.ffff.ffff.ffff.ffff.ffff.ffff.ffff.fffe.ffff.fc2f
> =p  ~(. fo prim)
> `@x`:(sum.p (exp.p x 3) 7)
0x71cad298f2cfc959e23d5b09def74add958bd72c9c55f8f14683019c926ccd0b
::  I got this version from here: https://stackoverflow.com/a/53480175
::
> `@x`(exp.p :(sum.p (exp.p x 3) 7) (div +(prim) 4))
0xd6eddba8206d458e1ec11219a28db8edf63d5c1a8d63befd739844c1122a20c0
> `@x`(mod :(add (pow x 3) 7) prim)
0xc1a9d327b2d943bd4664c2a5477341ff08e4d5d6a8fc3ae79d57eefd3ccd9397
```

Following the same thing in Javascript (and not using the `fo` operations) I get to this point with same results:

```
:: JS
> x.modPow(3, prime).add(7).mod(prime).toString(16)
'c1a9d327b2d943bd4664c2a5477341ff08e4d5d6a8fc3ae79d57eefd3ccd9397'
:: Hoon
> `@x`(mod (add (mod (pow x 3) prim) 7) prim)
0xc1a9d327b2d943bd4664c2a5477341ff08e4d5d6a8fc3ae79d57eefd3ccd9397
```
But then I either crash using normal `pow` or get a different result with `exp:fo`
```
:: JS
> x.modPow(3, prime).add(7).mod(prime).modPow( pIdent, prime ).toString(16)
'bcef10c989c7be8618ec41ceea16c0cb5365c6cc09c98514cf7ead8058bfd478'
:: Wrong result
> `@x`(exp.p (mod (add (mod (pow x 3) prim) 7) prim) (div +(prim) 4))
0xfb906dc75a416f3851d58eb2068b88c99fc9c9235f732651bda661ee13e5677b
:: Crash
> `@x`(mod (pow (mod (add (mod (pow x 3) prim) 7) prim) (div +(prim) 4)) prime)
bail: 3
crud: %belt event failed
bail: fail
  bar-stack
~[
  ~[/g/use/dojo/~zod/drum/hand /d //term/1]
  ~[/g/use/hood/~zod/out/~zod/dojo/drum/phat/~zod/dojo /d //term/1]
  ~[/d //term/1]
  ~[//term/1]
]
```

PPS: this will also need to check for the parity sign to select the proper root after y is calculated.

PPPS: next time I'll be looking at the source code before any trying anything; I was calling `+exp:fo` with the exponent and the base switched 

This is my fix to properly calculate `y`:

```
      ++  decompress-point
        ~/  %decompress-point
        |=  dat=@
        ^-  pont
        =+  x=(end 3 w dat)
        =+  y=(exp.p (div +(^p) 4) :(sum.p (exp.p 3 x) (pro.p a x) b))
        =+  s=(rsh 3 32 dat)
        ::  check parity
        ::
        =?  y  !=((sub s 2) (mod y 2))
          (sub ^p y)
        :-  x
        ?:  |(=(0x2 s) =(0x3 s))  y
        ~|  [`@ux`s `@ux`dat]
        !!
```
With this new decompress-point I can successfully derive new public keys from and extended-pubkey at different points of the tree (for xprv and xpub). 

I'll address this in a separate PR

**Comment by Fang-** [comment]
Woah, what an adventure. Nice work on figuring out the fix here! Would you want to open a PR with this fix (and maybe even some test cases for it), so that we can properly review this? (:

**Comment by yosoyubik** [comment]
> Would you want to open a PR with this fix

Yes, that's my plan, but this also depends on other changes to the bip32 library (to account for other types of extended public keys: regtest, tesnet... etc) I just need to wrap up a couple more things there and it should be ready. It'd be great to have that merged into master before I'm done with the other Bitcoin-related stuff; we've reached a functional POC state, now we just need to make it look pretty ;)

**Comment by joemfb** [comment]
Fixed in #2990
